### PR TITLE
[GDI32] Avoid invalid input in SetBkMode CORE-19656

### PR DESCRIPTION
**Pull request by turican0**
## Purpose

There is a bug in Midori installer 0.5.11 that probably causes overflow and bad rendering.
In Windows it is visible as FillRect( 0x89011a1f, 0x0019f87c, 0xabababab ), in ReactOS again as SetBkMode( 0xb5010fce, -1546139919 ).
Unfortunately this bug makes the text in ReactOS hard to read (no background).

Jira issue: [CORE-19656](https://jira.reactos.org/browse/CORE-19656)

## Proposed changes

After a number of attempts to get the same behavior, I haven't found a way to fix it without a hack so that the installer looks identical to Windows. So I added at least a fix for the incorrect value in SetBkMode. This shouldn't matter anywhere. It may not look the same as on Windows, but the text will at least be readable.

### CODE REVIEW

**Comment by turican0** [comment]
Midori 0.5.11 installer after fix (left - ReactOS, right - Windows 11):
image

File: win32ss/gdi/gdi32/objects/dc.c [hunk 1]
```diff
@@ -1057,6 +1057,8 @@ SetBkMode(
     _In_ HDC hdc,
     _In_ int iBkMode)
 {
+    if(iBkMode!=TRANSPARENT)iBkMode=OPAQUE;//fix for bad mode setting
```
Comment (oleg-dubinskiy) [inline win32ss/gdi/gdi32/objects/dc.c:1060]:
```suggestion
    /* Avoid bad mode setting */
    if (iBkMode != TRANSPARENT)
        iBkMode = OPAQUE;
```

**Review by oleg-dubinskiy — APPROVED** [review]
OK for the fix itself, but please improve a coding style as follows (see suggestion). It looks against or coding style guidelines. Thanks. :slightly_smiling_face:

**Comment by oleg-dubinskiy** [comment]
> Midori 0.5.11 installer after fix (left - ReactOS, right - Windows 11): image

But how it looks in Windows XP SP3/Windows Server 2003 SP2? Didn't you check it yet? ReactOS currently has a target Windows version (and is compatible to) Windows Server 2003 SP2, so ideally it should be compared with that first.

**Comment by turican0** [comment]
Díky, kódovací styl jsem opravil. Momentálně nemám Windows XP/2003 nainstalované, zítra to vyzkouším.

File: win32ss/gdi/gdi32/objects/dc.c [hunk 2]
```diff
@@ -1057,6 +1057,10 @@ SetBkMode(
     _In_ HDC hdc,
     _In_ int iBkMode)
 {
+    /* Avoid bad mode setting */
+    if(iBkMode!=TRANSPARENT)
+        iBkMode=OPAQUE;
```
Comment (oleg-dubinskiy) [inline win32ss/gdi/gdi32/objects/dc.c:1062]: Please add the spaces as well too, to make the code more readable.
```suggestion
    if(iBkMode != TRANSPARENT)
        iBkMode = OPAQUE;
```

File: win32ss/gdi/gdi32/objects/dc.c [hunk 3]
```diff
@@ -1057,6 +1057,10 @@ SetBkMode(
     _In_ HDC hdc,
     _In_ int iBkMode)
 {
+    /* Avoid bad mode setting */
+    if(iBkMode != TRANSPARENT)
```
Comment (binarymaster) [inline win32ss/gdi/gdi32/objects/dc.c:1061]:
```suggestion
    if (iBkMode != TRANSPARENT)
```

File: win32ss/gdi/gdi32/objects/dc.c [hunk 4]
```diff
@@ -1057,6 +1057,10 @@ SetBkMode(
     _In_ HDC hdc,
     _In_ int iBkMode)
 {
+    /* Avoid bad mode setting */
+    if (iBkMode != TRANSPARENT)
+        iBkMode = OPAQUE;
+	
     PDC_ATTR pdcattr;
     INT iOldMode;
```
Comment (binarymaster) [inline win32ss/gdi/gdi32/objects/dc.c:1066]: Suggestion for keeping the code C-compliant
```suggestion
    PDC_ATTR pdcattr;
    INT iOldMode;

    /* Avoid bad mode setting */
    if (iBkMode != TRANSPARENT)
        iBkMode = OPAQUE;

```

**Review by HBelusca — APPROVED** [review]
let's say OK for now; I have in a branch more extensive handling for this (here and elsewhere) but I would also provide some apitests for it.

File: win32ss/gdi/gdi32/objects/dc.c [hunk 5]
```diff
@@ -1060,6 +1060,10 @@ SetBkMode(
...
 
+    /* Avoid bad mode setting */
+    if (iBkMode != TRANSPARENT)
+        iBkMode = OPAQUE;
```
[the remark this answers is not in the document]
Comment (turican0) [reply win32ss/gdi/gdi32/objects/dc.c:1065]: &[SergeGautherie](https://github.com/SergeGautherie)You mean something like:
{code:c++}
    if((iBkMode != TRANSPARENT)&(iBkMode != TRANSPARENT))
    {
        iBkMode = OPAQUE;
        ERR("SetBkMode - incorrect value");
    }
{code}

[the remark this answers is not in the document]
Comment (turican0) [reply win32ss/gdi/gdi32/objects/dc.c:1065]: For the Midori installation it behaves the same way(I tried this code).For me it doesn't really matter unless we have more test cases, but at least the logging is worth adding.:
```
    if((iBkMode != TRANSPARENT)&(iBkMode != TRANSPARENT))
    {
        DbgPrint ("ERR: SetBkMode - incorrect value");
        return 0;
    }
```
image

[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/gdi/gdi32/objects/dc.c:1065]: Yeah, but use `ERR` macro instead of `DbgPrint`.

[the remark this answers is not in the document]
Comment (turican0) [reply win32ss/gdi/gdi32/objects/dc.c:1065]: I tried that, but it wouldn't compile:
../win32ss/gdi/gdi32/objects/dc.c:1066:9: error: implicit declaration of function 'ERR'; did you mean 'ERR_'? [-Werror=implicit-function-declaration]
         ERR("SetBkMode - incorrect value");
         ^~~
         ERR_
cc1.exe: all warnings being treated as errors
ninja: build stopped: subcommand failed.

[the remark this answers is not in the document]
Comment (oleg-dubinskiy) [reply win32ss/gdi/gdi32/objects/dc.c:1065]: Ah, indeed, my mistake. This file has no Wine debug channel, so `DPRINT1` should work instead. It's our macro around `DbgPrint`.

**Comment by turican0** [comment]
I checked the behavior under WinXP, it is similar to Win11. Same rendering and bad SetBkMode.
image
image

**Comment by turican0** [comment]
And I changed the code to the version with the Debug message.

**Review by HBelusca — APPROVED** [review]

File: win32ss/gdi/gdi32/objects/dc.c [hunk 6]
```diff
@@ -1060,6 +1060,13 @@ SetBkMode(
     PDC_ATTR pdcattr;
     INT iOldMode;
 
+    /* Avoid bad mode setting */
+    if((iBkMode != TRANSPARENT) & (iBkMode != TRANSPARENT))
```
Comment (Doug-Lyons) [inline win32ss/gdi/gdi32/objects/dc.c:1064]: Why are you checking the exact same condition twice? I suspect that you intended instead:
```
    if ((iBkMode != TRANSPARENT) & (iBkMode != OPAQUE))
```
Also, note <space> after the ```if```.

**Review by Doug-Lyons — APPROVED** [review]
LGTM. Thanks for your great trouble-shooting on this issue.

File: win32ss/gdi/gdi32/objects/dc.c [hunk 7]
```diff
@@ -1060,6 +1060,13 @@ SetBkMode(
     PDC_ATTR pdcattr;
     INT iOldMode;
 
+    /* Avoid bad mode setting */
+    if ((iBkMode != TRANSPARENT) && (iBkMode != OPAQUE))
+    {
+        DPRINT1("SetBkMode - incorrect value");
```
Comment (binarymaster) [inline win32ss/gdi/gdi32/objects/dc.c:1066]:
```suggestion
        DPRINT1("SetBkMode: Incorrect value\n");
```

**Comment by binarymaster** [comment]
@turican0 this one is ready for merge, please update your GitHub profile name with correct diacritics, [as in here](https://github.com/reactos/reactos/pull/7100/files#diff-377583c24f811846cd94caab3ea73598157ef88e9a166e55b0e508e0ab7dc346R6).

**Comment by turican0** [comment]
I added an accent to the name in the profile.

**Pull request merged**